### DISCUSSION TITLE
How to directly set the value of a ForeignKey in a Dataclass mixin

### DISCUSSION THREAD

**Discussion by johnniemorrow** [sqlalchemy/sqlalchemy#9856] [Usage Questions] [answered]
Hi!

I have a model with an EventMixin class used by other model classes. The EventMixin class has some foreign key relationships defined using @declared_attr. Most of the time I already know the value of those foreign key ids when I'm creating an Event so I can just populate the foreign key id fields directly instead of doing a select/get to look them up.

This was working fine with regular classes. Then I tried updating my model to use dataclasses but directly populating the mixin foreign key id fields isn't working for me any more. I'm not sure if I'm doing something wrong, or maybe dataclasses+mixins are not meant to be used in the same way.

Below is a pared-down version of the model - the most important parts are that Event has a direct FK relationship to EventFile, and an inherited FK relationship to Location (from EventMixin). When I create an Event, setting the file_id FK works, but setting the location_id FK doesn't.

```python
from typing import Optional
import sqlalchemy
from sqlalchemy import ForeignKey

from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import MappedAsDataclass
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class EventFile(Base):
    __tablename__ = "event_file"
    id: Mapped[int] = mapped_column(primary_key=True, init=True, default=None)
    name: Mapped[str] = mapped_column(default=None)


class Location(Base):
    __tablename__ = "location"
    id: Mapped[int] = mapped_column(primary_key=True, default=None)
    city: Mapped[str] = mapped_column(default=None)
    state: Mapped[str] = mapped_column(default=None)


class EventMixin(MappedAsDataclass): # also tried: class EventMixin:
    id: Mapped[int] = mapped_column(primary_key=True, init=True, default=None)
    name: Mapped[str] = mapped_column(default=None)
    location_id: Mapped[Optional[int]] = mapped_column(ForeignKey("location.id"), default=None) # also tried init=False)
    @declared_attr
    def location(cls) -> Mapped["Location"]:
        return relationship(Location, default=None)


class Event(EventMixin, Base):
    __tablename__ = "event"
    file_id: Mapped[Optional[int]] = mapped_column(ForeignKey("event_file.id"), default=None)
    file = relationship(EventFile, default=None)


if __name__ == "__main__":
    db_url = "postgresql+psycopg://postgres:secret@127.0.0.1:5432/postgres"
    engine = sqlalchemy.create_engine(db_url, echo=False)
    Base.metadata.create_all(bind=engine)

    with sqlalchemy.orm.Session(engine) as session:
        session.add(Location(id=1, city="New York", state="NY"))
        session.add(Location(id=2, city="Chicago", state="IL"))
        session.add(EventFile(id=1, name="event file"))
        session.commit()

        session.add(Event(id=1, name="event", location_id=1, file_id=1))
        # also tried, with init=False above
        #event = Event(id=1, name="event", file_id=1)
        #event.location_id = 1
        #session.add(event)

        engine.echo = True
        session.commit()
```

I ran it in a venv with the dependencies below, and Postgres running in Docker: 
```sh
$ (venv) cat requirements.txt
psycopg==3.1.9
SQLAlchemy==2.0.15

$ (venv) docker run --name postgres-db -e POSTGRES_PASSWORD=secret -d -p 5432:5432 postgres:13.9
```

When it runs, I can see from the echoed SQL that it's passing in the file id, but not the location id:
```sh
2023-05-29 20:42:11,819 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-05-29 20:42:11,821 INFO sqlalchemy.engine.Engine INSERT INTO event (file_id, id, name, location_id) VALUES (%(file_id)s::INTEGER, %(id)s::INTEGER, %(name)s::VARCHAR, %(location_id)s::INTEGER)
2023-05-29 20:42:11,821 INFO sqlalchemy.engine.Engine [generated in 0.00039s] {'file_id': 1, 'id': 1, 'name': 'event', 'location_id': None}
2023-05-29 20:42:11,825 INFO sqlalchemy.engine.Engine COMMIT
```
...and when I check the database, I can see that the file fk is set but the location one isn't:

```sh
$ (venv) PGPASSWORD=secret psql -h localhost -p 5432 -U postgres

postgres=# \d event
                                    Table "public.event"
   Column    |       Type        | Collation | Nullable |              Default              
-------------+-------------------+-----------+----------+-----------------------------------
 file_id     | integer           |           |          | 
 id          | integer           |           | not null | nextval('event_id_seq'::regclass)
 name        | character varying |           | not null | 
 location_id | integer           |           |          | 
Indexes:
    "event_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "event_file_id_fkey" FOREIGN KEY (file_id) REFERENCES event_file(id)
    "event_location_id_fkey" FOREIGN KEY (location_id) REFERENCES location(id)

postgres=# select * from event;
 file_id | id | name  | location_id 
---------+----+-------+-------------
       1 |  1 | event |            
(1 row)
```
I saw [this question](https://github.com/sqlalchemy/sqlalchemy/discussions/9383) and thought that might be it, and tried adding init=False to the location_id declaration and then populating the location_id field after creating the Event, but still the same:
```python
        # event = Event(id=1, name="event", location_id=1, file_id=1)
        event = Event(id=1, name="event", file_id=1)
        event.location_id = 1
        session.add(event)
```

Is there a way to set the value of a foreign key field from a mixin directly? I have tried getting the location objects by key, doing `event.location = session.get(Location, 1)` and that does work, but in my real version I have a good few foreign keys and the extra lookups have a performance impact.

Thanks!
John

**Comment by zzzeek** [comment, marked as answer]
the problem is your location relationship also setting a default of None:

```py
class EventMixin(MappedAsDataclass): # also tried: class EventMixin:
    id: Mapped[int] = mapped_column(primary_key=True, init=True, default=None)
    name: Mapped[str] = mapped_column(default=None)
    location_id: Mapped[Optional[int]] = mapped_column(ForeignKey("location.id"), default=None) # also tried init=False)
    @declared_attr
    def location(cls) -> Mapped["Location"]:
        return relationship(Location, default=None)
```

this creates the constructor as follows:


```py
def __init__(self, id=None, name=None, location_id=None, location=None)
```

meaning creating an Event is as though you did this using non-dataclasses:

```py
e = Event()
e.id = 1
e.name = "event"
e.location_id = 1
e.location = None   # this supersedes location_id

```

this issue is a new thing with dataclasses and a similar thing was mentioned at #9383 - it then links to a pretty big discussion about this dataclasses gotcha in general at https://stackoverflow.com/questions/55099243/python3-dataclass-with-kwargsasterisk .

for us, we might need to add some kind of magic variable as mentioned in the comments at #9383.

**Reply by johnniemorrow** [reply]
Thanks very much, Mike! I didn't think of location also being in the init as well as location_id! I just added init=False to the location relationship and now that allows me to specify the location_id without it being overwritten by location - that's great!

I wonder why the same thing didn't happen with the file/file_id parameter pair - there I'm specifying the file_id but in that case it's not getting overwritten by the file relationship. Maybe that ends up as:
```
e.id = 1
...
e.file = None
e.file_id = 1
```
...and it's just lucky ordering in that case. I've just added an init=False to the file one too, and I guess for any of these pairs I should always switch init off for one of them so the construction is predictable.

Thanks again!

**Comment by zzzeek** [comment]
Here's a generalized recipe for ignoring m2o relationship as None for the 2.0 series until #12168 is available: https://github.com/sqlalchemy/sqlalchemy/issues/12168#issuecomment-2892092715
